src/emscripten-optimizer/simple_ast.h

```diff
@@ -537,7 +537,266 @@ void traverseFunctions(Ref ast, std::function<void (Ref)> visit);
 // JS printing support
 
 struct JSPrinter {
-  static char* numToString(double d, bool finalize=true) {
+  bool pretty, finalize;
+
+  char *buffer;
+  size_t size, used;
+
+  int indent;
+  bool possibleSpace; // add a space to separate identifiers
+
+  Ref ast;
+
+  JSPrinter(bool pretty_, bool finalize_, Ref ast_) : pretty(pretty_), finalize(finalize_), buffer(0), size(0), used(0), indent(0), possibleSpace(false), ast(ast_) {}
+
+  void printAst() {
+    print(ast);
+    buffer[used] = 0;
+  }
+
+  // Utils
+
+  void ensure(int safety=100) {
+    if (size < used + safety) {
+      size = std::max((size_t)1024, size * 2) + safety;
+      if (!buffer) {
+        buffer = (char*)malloc(size);
+        if (!buffer) {
+          printf("Out of memory allocating %zd bytes for output buffer!", size);
+          abort();
+        }
+      } else {
+        char *buf = (char*)realloc(buffer, size);
+        if (!buf) {
+          free(buffer);
+          printf("Out of memory allocating %zd bytes for output buffer!", size);
+          abort();
+        }
+        buffer = buf;
+      }
+    }
+  }
+
+  void emit(char c) {
+    maybeSpace(c);
+    if (!pretty && c == '}' && buffer[used-1] == ';') used--; // optimize ;} into }, the ; is not separating anything
+    ensure(1);
+    buffer[used++] = c;
+  }
+
+  void emit(const char *s) {
+    maybeSpace(*s);
+    int len = strlen(s);
+    ensure(len+1);
+    strncpy(buffer + used, s, len+1);
+    used += len;
+  }
+
+  void newline() {
+    if (!pretty) return;
+    emit('\n');
+    for (int i = 0; i < indent; i++) emit(' ');
+  }
+
+  void space() {
+    if (pretty) emit(' ');
+  }
+
+  void safeSpace() {
+    if (pretty) emit(' ');
+    else possibleSpace = true;
+  }
+
+  void maybeSpace(char s) {
+    if (possibleSpace) {
+      possibleSpace = false;
+      if (isIdentPart(s)) emit(' ');
+    }
+  }
+
+  void print(Ref node) {
+    ensure();
+    if (node->isString()) {
+      printName(node);
+      return;
+    }
+    if (node->isNumber()) {
+      printNum(node);
+      return;
+    }
+    if (node->isAssign()) {
+      printAssign(node);
```

Comment:
This code had previously been deleted, so I just brought it back exactly how it was. But yeah I don't see why there wouldn't be a return here.